// @zent/design/es/Design.js
import _extends from 'babel-runtime/helpers/extends';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _isFunction from 'lodash/isFunction';
import _defaultTo from 'lodash/defaultTo';
import _isUndefined from 'lodash/isUndefined';
import _isEmpty from 'lodash/isEmpty';
import _findIndex from 'lodash/findIndex';
import _find from 'lodash/find';
import _assign from 'lodash/assign';

var _class, _temp, _initialiseProps;

/* eslint-disable no-script-url */

/**
 * 设计文档
 *
 * 预览
 * `DesignPreview` 组件是整个预览块的包裹层，负责渲染左侧预览的框架。`DesignPreview` 和 `config`
 * 子组件是相关的，`config` 组件是知道 `DesignPreview` 的存在的；而 `DesignPreview` 的渲染是
 * 根据 `config` 生成的数据进行的。
 * ⚠️注意：`config` 自身有相应的负责渲染预览的模块，这个和 `DesignPreview` 不冲突，可以理解成
 * `config` 可以控制一些预览界面的全局样式。

 * 预览界面中按模块分成很多区域，每个区域是一个 `DesignPreviewItem`，默认的 `DesignPreviewItem`
 * 实现可以由外部覆盖。负责每个区域的事件交互的是另一个组件 `DesignPreviewController`，这个组件
 * 负责处理添加、删除、编辑、选中以及拖拽操作，`DesignPreviewController` 的实现也是可以由外部覆盖的。
 * ⚠️注意：重写的时候所有交互都需要再这个组件里面处理。`DesignPreviewController` 内部会渲染该区域
 * 对应组件的预览模块，预览模块有两个参数：`value` 和 `design`。`value` 是当前的值，`design` 是
 *  `Design` 组件提供的一些操作，一般用不到。

 * 编辑
 * `DesignEditorItem` 是每个区域对应的编辑区域，这个区域的显示隐藏由 `Design` 控制。`DesignEditorItem`
 * 可以由外部覆盖重写。

 * `DesignEditorAddComponent` 这个组件负责枚举所有**可以添加的组件**，暂不支持由外部自定义组件实现。

 * `DesignEditor` 是所有编辑组件的基类，这个类提供了一些常用的方法（例如 `onChange` 事件的处理函数），
 * 在子类里面可以直接使用。
 */

import React, { PureComponent } from 'react';
import { findDOMNode } from 'react-dom';

import cx from 'classnames';

import Alert from 'zent/es/alert';
import * as storage from 'zent/es/utils/storage';

import uuid from './utils/uuid';

import DesignPreview from './preview/DesignPreview';
import { getDesignType, isExpectedDesignType, serializeDesignType } from './utils/design-type';
import LazyMap from './utils/LazyMap';
import { ADD_COMPONENT_OVERLAY_POSITION } from './constants';

var UUID_KEY = '__zent-design-uuid__';
var CACHE_KEY = '__zent-design-cache-storage__';

var hasValidateError = function hasValidateError(v) {
  return !_isEmpty(v[Object.keys(v)[0]]);
};

var Design = (_temp = _class = function (_PureComponent) {
  _inherits(Design, _PureComponent);

  function Design(props) {
    _classCallCheck(this, Design);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _initialiseProps.call(_this);

    var value = props.value,
        defaultSelectedIndex = props.defaultSelectedIndex;


    _this.validateCacheProps(props);

    tagValuesWithUUID(value);

    var safeValueIndex = getSafeSelectedValueIndex(defaultSelectedIndex, value);
    var selectedValue = value[safeValueIndex];

    _this.state = {
      // 当前选中的组件对应的 UUID
      selectedUUID: _this.getUUIDFromValue(selectedValue),

      // 每个组件当前已经添加的个数
      componentInstanceCount: makeInstanceCountMapFromValue(props.value, props.components),

      // 外面没传的时候用 state 上的 settings
      settings: {},

      // 是否显示添加组件的浮层
      showAddComponentOverlay: false,

      // 添加组件浮层的位置
      addComponentOverlayPosition: ADD_COMPONENT_OVERLAY_POSITION.UNKNOWN,

      // 可添加的组件列表
      appendableComponents: [],

      // 当前所有组件的 validation 信息
      // key 是 value 的 UUID
      validations: {},

      // 是否强制显示错误
      showError: false,

      // 是否显示从缓存中恢复的提示
      showRestoreFromCache: false,

      // 当 preview 很长时，为了对齐 preview 底部需要的额外空间
      bottomGap: 0
    };
    return _this;
  }

  Design.prototype.render = function render() {
    var _props = this.props,
        className = _props.className,
        prefix = _props.prefix,
        preview = _props.preview,
        cacheRestoreMessage = _props.cacheRestoreMessage,
        children = _props.children;
    var _state = this.state,
        showRestoreFromCache = _state.showRestoreFromCache,
        bottomGap = _state.bottomGap;

    var cls = cx(prefix + '-design', className);

    return React.createElement(
      'div',
      { className: cls, style: { paddingBottom: bottomGap } },
      showRestoreFromCache && React.createElement(
        Alert,
        {
          className: prefix + '-design__restore-cache-alert',
          closable: true,
          onClose: this.onRestoreCacheAlertClose,
          type: 'warning'
        },
        cacheRestoreMessage,
        React.createElement(
          'a',
          {
            className: prefix + '-design__restore-cache-alert-use',
            onClick: this.restoreCache,
            href: 'javascript:void(0);'
          },
          '\u4F7F\u7528'
        )
      ),
      this.renderPreview(preview),
      children
    );
  };

  Design.prototype.componentWillMount = function componentWillMount() {
    this.cacheAppendableComponents(this.props.components);
  };

  Design.prototype.componentDidMount = function componentDidMount() {
    this.setupBeforeUnloadHook();
    this.checkCache();
  };

  Design.prototype.componentDidUpdate = function componentDidUpdate() {
    this.setupBeforeUnloadHook();
  };

  Design.prototype.componentWillUnmount = function componentWillUnmount() {
    this.uninstallBeforeUnloadHook();
  };

  Design.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    this.validateCacheProps(nextProps);

    var shouldUpdateInstanceCountMap = false;

    if (nextProps.value !== this.props.value) {
      tagValuesWithUUID(nextProps.value);
      shouldUpdateInstanceCountMap = true;
    }

    if (nextProps.components !== this.props.components) {
      this.cacheAppendableComponents(nextProps.components);
      shouldUpdateInstanceCountMap = true;
    }

    // 如果当前没有选中的并且 value 或者 defaultSelectedIndex 改变的话
    // 重新尝试设置默认值
    if (!this.hasSelected() && (nextProps.defaultSelectedIndex !== this.props.defaultSelectedIndex || nextProps.value !== this.props.value)) {
      var value = nextProps.value,
          defaultSelectedIndex = nextProps.defaultSelectedIndex;

      this.selectByIndex(defaultSelectedIndex, value);
    }

    if (shouldUpdateInstanceCountMap) {
      this.setState({
        componentInstanceCount: makeInstanceCountMapFromValue(nextProps.value, nextProps.components)
      });
    }
  };

  Design.prototype.cacheAppendableComponents = function cacheAppendableComponents(components) {
    this.setState({
      appendableComponents: components.filter(function (c) {
        return c.appendable === undefined || c.appendable;
      })
    });
  };

  Design.prototype.renderPreview = function renderPreview(preview) {
    var _props2 = this.props,
        components = _props2.components,
        prefix = _props2.prefix,
        value = _props2.value,
        disabled = _props2.disabled,
        settings = _props2.settings,
        previewFooter = _props2.previewFooter,
        globalConfig = _props2.globalConfig;
    var _state2 = this.state,
        selectedUUID = _state2.selectedUUID,
        appendableComponents = _state2.appendableComponents,
        showAddComponentOverlay = _state2.showAddComponentOverlay,
        addComponentOverlayPosition = _state2.addComponentOverlayPosition,
        validations = _state2.validations,
        showError = _state2.showError,
        managedSettings = _state2.settings,
        componentInstanceCount = _state2.componentInstanceCount;


    return React.createElement(preview, {
      prefix: prefix,
      components: components,
      value: value,
      validations: validations,
      showError: showError,
      settings: settings || managedSettings,
      onSettingsChange: this.onSettingsChange,
      footer: previewFooter,
      componentInstanceCount: componentInstanceCount,
      onComponentValueChange: this.onComponentValueChange,
      onAddComponent: this.onAdd,
      appendableComponents: appendableComponents,
      selectedUUID: selectedUUID,
      getUUIDFromValue: this.getUUIDFromValue,
      showAddComponentOverlay: showAddComponentOverlay,
      addComponentOverlayPosition: addComponentOverlayPosition,
      onAdd: this.onShowAddComponentOverlay,
      onEdit: this.onShowEditComponentOverlay,
      onSelect: this.onSelect,
      onMove: this.onMove,
      onDelete: this.onDelete,
      design: this.design,
      globalConfig: globalConfig,
      disabled: disabled,
      ref: this.savePreview
    });
  };

  // 打开右侧添加新组件的弹层


  // 编辑一个已有组件


  // 选中一个组件


  // 添加一个新组件


  // 删除一个组件


  // 验证所有组件，如果有错误选中并跳转到第一个有错误的组件。
  // 如果没有错误，Promise resolve；如果有错误，Promise reject。
  // reject 的是个数组，
  // [
  //   { '508516bf-d3e5-40a5-812e-834d3dee1d54': {} },
  //   { 'c7c72599-2ac5-41bb-9ba0-45e8178ff5a6': { content: '请填写公告内容' } }
  // ]


  // 保存数据后请调用这个函数通知组件数据已经保存


  Design.prototype.toggleEditOrAdd = function toggleEditOrAdd(component, showAdd) {
    var addPosition = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : ADD_COMPONENT_OVERLAY_POSITION.UNKNOWN;
    var _state3 = this.state,
        showAddComponentOverlay = _state3.showAddComponentOverlay,
        addComponentOverlayPosition = _state3.addComponentOverlayPosition;

    var id = this.getUUIDFromValue(component);

    if (this.isSelected(component) && showAddComponentOverlay === showAdd && addPosition === addComponentOverlayPosition) {
      return;
    }

    this.setState({
      selectedUUID: id,
      showAddComponentOverlay: showAdd,
      addComponentOverlayPosition: addPosition
    });
    this.adjustHeight();
  };

  Design.prototype.getUUIDFromValue = function getUUIDFromValue(value) {
    return value && value[UUID_KEY];
  };

  Design.prototype.setUUIDForValue = function setUUIDForValue(value, id) {
    if (value) {
      value[UUID_KEY] = id;
    }

    return value;
  };

  // 滚动到第一个有错误的组件
  Design.prototype.scrollToPreviewItem = function scrollToPreviewItem(id) {
    if (this.preview) {
      var _props3 = this.props,
          scrollTopOffset = _props3.scrollTopOffset,
          scrollLeftOffset = _props3.scrollLeftOffset;

      this.preview.scrollToItem && this.preview.scrollToItem(id, {
        top: scrollTopOffset,
        left: scrollLeftOffset
      });
    }
  };

  // 调整 Design 的高度，因为 editor 是 position: absolute 的，所以需要动态的更新
  // 实际并未改变高度，而是设置了margin/padding


  // 调用 onChange 的统一入口，用于处理一些需要知道有没有修改过值的情况
  Design.prototype.trackValueChange = function trackValueChange(newValue) {
    var writeCache = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
    var onChange = this.props.onChange;

    onChange(newValue);

    if (!this._dirty) {
      this._dirty = true;
    }

    if (writeCache) {
      this.writeCache(newValue);
    }

    this.adjustHeight();
  };

  Design.prototype.setupBeforeUnloadHook = function setupBeforeUnloadHook() {
    var confirmUnsavedLeave = this.props.confirmUnsavedLeave;


    if (this._hasBeforeUnloadHook || !confirmUnsavedLeave) {
      return;
    }

    window.addEventListener('beforeunload', this.onBeforeWindowUnload);
    this._hasBeforeUnloadHook = true;
  };

  Design.prototype.uninstallBeforeUnloadHook = function uninstallBeforeUnloadHook() {
    window.removeEventListener('beforeunload', this.onBeforeWindowUnload);
    this._hasBeforeUnloadHook = false;
  };

  // 缓存相关的函数
  Design.prototype.validateCacheProps = function validateCacheProps(props) {
    props = props || this.props;
    var _props4 = props,
        cache = _props4.cache,
        cacheId = _props4.cacheId;

    if (cache && !cacheId) {
      throw new Error('Design: cacheId is required when cache is on');
    }
  };

  Design.prototype.checkCache = function checkCache() {
    var cache = this.props.cache;


    if (cache) {
      var cachedValue = this.readCache();

      if (cachedValue !== storage.NOT_FOUND) {
        this.setState({
          showRestoreFromCache: true
        });
      }
    }
  };

  Design.prototype.readCache = function readCache() {
    var cache = this.props.cache;

    if (!cache) {
      return storage.NOT_FOUND;
    }

    var cacheId = this.props.cacheId;

    return storage.read(CACHE_KEY, cacheId);
  };

  Design.prototype.writeCache = function writeCache(value) {
    var cache = this.props.cache;

    if (!cache) {
      return false;
    }

    var cacheId = this.props.cacheId;

    return storage.write(CACHE_KEY, cacheId, value);
  };

  Design.prototype.removeCache = function removeCache() {
    // 这个函数不需要检查有没有开启缓存，强制清除
    var cacheId = this.props.cacheId;

    return storage.write(CACHE_KEY, cacheId, undefined);
  };

  // 关闭提示，但是不清楚缓存


  // 恢复缓存的数据并删除缓存


  // Dummy method to make Design and DesignWithDnd compatible at source code level
  Design.prototype.getDecoratedComponentInstance = function getDecoratedComponentInstance() {
    return this;
  };

  // Actions on design


  return Design;
}(PureComponent), _class.defaultProps = {
  preview: DesignPreview,
  value: [],
  defaultSelectedIndex: -1,
  globalConfig: {},
  confirmUnsavedLeave: true,
  cache: false,
  cacheRestoreMessage: '提示：在浏览器中发现未提交的内容，是否使用该内容替换当前内容？',
  scrollTopOffset: -10,
  scrollLeftOffset: -10,
  prefix: 'zent'
}, _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.onSettingsChange = function (value) {
    var _props5 = _this2.props,
        settings = _props5.settings,
        onSettingsChange = _props5.onSettingsChange;

    var onSettingsChangeExists = _isFunction(onSettingsChange);

    if (settings && !onSettingsChangeExists) {
      throw new Error('Expects onSettingsChange to be a function');
    }

    if (settings && onSettingsChangeExists) {
      onSettingsChange(_extends({}, settings, value));
    }

    if (!settings) {
      _this2.setState({
        settings: _extends({}, _this2.state.settings, value)
      });
    }
  };

  this.onComponentValueChange = function (identity) {
    return function (diff) {
      var _assign2;

      var replace = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var value = _this2.props.value;

      var newComponentValue = replace ? _assign((_assign2 = {}, _assign2[UUID_KEY] = _this2.getUUIDFromValue(identity), _assign2), diff) : _assign({}, identity, diff);
      var newValue = value.map(function (v) {
        return v === identity ? newComponentValue : v;
      });
      var changedProps = Object.keys(diff);

      _this2.trackValueChange(newValue);
      _this2.validateComponentValue(newComponentValue, identity, changedProps).then(function (errors) {
        var _this2$setValidation;

        var id = _this2.getUUIDFromValue(newComponentValue);
        _this2.setValidation((_this2$setValidation = {}, _this2$setValidation[id] = errors, _this2$setValidation));
      });
    };
  };

  this.validateComponentValue = function (value, prevValue, changedProps) {
    var type = value.type;
    var components = _this2.props.components;

    var comp = _find(components, function (c) {
      return isExpectedDesignType(c, type);
    });
    var validate = comp.editor.validate;

    var p = validate(value, prevValue, changedProps);

    return p;
  };

  this.onShowAddComponentOverlay = function (component, addPosition) {
    _this2.toggleEditOrAdd(component, true, addPosition);
  };

  this.onShowEditComponentOverlay = function (component) {
    _this2.toggleEditOrAdd(component, false);

    // 将当前组件滚动到顶部
    var id = _this2.getUUIDFromValue(component);
    _this2.scrollToPreviewItem(id);
  };

  this.onSelect = function (component) {
    var id = _this2.getUUIDFromValue(component);
    var showAddComponentOverlay = _this2.state.showAddComponentOverlay;


    if (_this2.isSelected(component) && !showAddComponentOverlay) {
      return;
    }

    _this2.setState({
      selectedUUID: id,
      showAddComponentOverlay: false
    });

    _this2.adjustHeight();
  };

  this.onAdd = function (component, fromSelected) {
    var _props6 = _this2.props,
        value = _props6.value,
        settings = _props6.settings,
        globalConfig = _props6.globalConfig;
    var editor = component.editor,
        defaultType = component.defaultType;

    var instance = editor.getInitialValue({
      settings: settings,
      globalConfig: globalConfig
    });
    instance.type = getDesignType(editor, defaultType);
    var id = uuid();
    _this2.setUUIDForValue(instance, id);

    /**
     * 添加有两种来源：底部区域或者弹层。
     * 如果来自底部的话，就在当前数组最后加；如果来自弹层就在当前选中的那个组件后面加
     */
    var newValue = void 0;
    if (fromSelected) {
      var _findIndex2;

      newValue = value.slice();
      var addComponentOverlayPosition = _this2.state.addComponentOverlayPosition;
      var selectedUUID = _this2.state.selectedUUID;

      var selectedIndex = _findIndex(value, (_findIndex2 = {}, _findIndex2[UUID_KEY] = selectedUUID, _findIndex2));

      // 两种位置，插入到当前选中的前面或者后面
      var delta = addComponentOverlayPosition === ADD_COMPONENT_OVERLAY_POSITION.TOP ? 0 : 1;
      newValue.splice(selectedIndex + delta, 0, instance);
    } else {
      newValue = value.concat(instance);
    }

    _this2.trackValueChange(newValue);
    _this2.onSelect(instance);
  };

  this.onDelete = function (component) {
    var _props7 = _this2.props,
        value = _props7.value,
        components = _props7.components;

    var nextIndex = -1;
    var newValue = value.filter(function (v, idx) {
      var skip = v !== component;
      if (!skip) {
        nextIndex = idx - 1;
      }
      return skip;
    });

    var newState = {
      showAddComponentOverlay: false
    };

    // 删除选中项目后默认选中前一项可选的，如果不存在则往后找一个可选项
    var componentUUID = _this2.getUUIDFromValue(component);
    if (componentUUID === _this2.state.selectedUUID) {
      var nextSelectedValue = findFirstEditableSibling(newValue, components, nextIndex);
      var nextUUID = _this2.getUUIDFromValue(nextSelectedValue);
      newState.selectedUUID = nextUUID;
    }

    _this2.trackValueChange(newValue);
    _this2.setState(newState);

    _this2.adjustHeight();
  };

  this.onMove = function (fromIndex, toIndex) {
    if (fromIndex === toIndex) {
      return;
    }

    var _props8 = _this2.props,
        value = _props8.value,
        components = _props8.components;

    var newValue = [];
    var tmp = void 0;

    /**
     * 这个算法不是仅仅交换两个位置的节点，所有中间节点都需要移位
     * 需要考虑数组中间有不可拖拽节点的情况，这种情况下 fromIndex, toIndex 的值是不包括这些节点的
     * 例如 [1, 0, 0, 1, 0, 0, 1]: fromIndex = 0, toIndex = 1 表示移动第一个和第二个 1。
     */
    var passedFromIndex = false;
    var passedToIndex = false;

    if (fromIndex < toIndex) {
      var _loop = function _loop(i, _dragableIndex) {
        var val = value[i];

        var comp = _find(components, function (c) {
          return isExpectedDesignType(c, val.type);
        });
        var dragable = comp && _defaultTo(comp.dragable, true);
        if (dragable) {
          _dragableIndex++;
        }

        /* Invariant: Each step copies one value, except one copies 2 and another doesn't copy */
        if (_dragableIndex === fromIndex && !passedFromIndex) {
          tmp = val;
          passedFromIndex = true;
        } else if (_dragableIndex < toIndex && passedFromIndex) {
          newValue[i - 1] = val;
        } else if (_dragableIndex === toIndex && !passedToIndex) {
          newValue[i - 1] = val;
          newValue[i] = tmp;
          passedToIndex = true;
        } else {
          newValue[i] = val;
        }
        dragableIndex = _dragableIndex;
      };

      for (var i = 0, dragableIndex = -1; i < value.length; i++) {
        _loop(i, dragableIndex);
      }
    } else {
      var toInsetIndex = void 0;

      var _loop2 = function _loop2(i, _dragableIndex3) {
        var val = value[i];

        var comp = _find(components, function (c) {
          return isExpectedDesignType(c, val.type);
        });
        var dragable = comp && _defaultTo(comp.dragable, true);
        if (dragable) {
          _dragableIndex3++;
        }

        /* Invariant: each step copies one value */
        if (_dragableIndex3 === toIndex && !passedToIndex) {
          toInsetIndex = i;
          newValue[i + 1] = val;
          passedToIndex = true;
        } else if (_dragableIndex3 < fromIndex && passedToIndex) {
          newValue[i + 1] = val;
        } else if (_dragableIndex3 === fromIndex && !passedFromIndex) {
          newValue[toInsetIndex] = val;
          passedFromIndex = true;
        } else {
          newValue[i] = val;
        }
        _dragableIndex2 = _dragableIndex3;
      };

      for (var i = 0, _dragableIndex2 = -1; i < value.length; i++) {
        _loop2(i, _dragableIndex2);
      }
    }

    _this2.trackValueChange(newValue);
  };

  this.setValidation = function (validation) {
    _this2.setState({
      validations: _assign({}, _this2.state.validations, validation)
    });

    _this2.adjustHeight();
  };

  this.validate = function () {
    var _props9 = _this2.props,
        value = _props9.value,
        components = _props9.components;


    return new Promise(function (resolve, reject) {
      return Promise.all(value.map(function (v) {
        var id = _this2.getUUIDFromValue(v);
        var type = v.type;

        var comp = _find(components, function (c) {
          return isExpectedDesignType(c, type);
        });
        // 假如组件设置了 editable: false，不处罚校验
        if (!_defaultTo(comp.editable, true)) {
          var _Promise$resolve;

          return Promise.resolve((_Promise$resolve = {}, _Promise$resolve[id] = {}, _Promise$resolve));
        }

        return _this2.validateComponentValue(v, v, {}).then(function (errors) {
          var _ref;

          return _ref = {}, _ref[id] = errors, _ref;
        });
      })).then(function (validationList) {
        var validations = _assign.apply(undefined, [{}].concat(validationList));

        _this2.setState({
          showError: true,
          validations: validations
        }, function () {
          // 跳转到第一个有错误的组件
          var firstError = _find(validationList, hasValidateError);

          if (firstError) {
            var id = Object.keys(firstError)[0];
            _this2.scrollToPreviewItem(id);

            // 选中第一个有错误的组件
            _this2.setState({
              selectedUUID: id,
              showAddComponentOverlay: false
            });
          }

          _this2.adjustHeight();
        });

        // 过滤所有错误信息，将数组合并为一个对象，key 是每个组件的 id
        var validationErrors = validationList.filter(hasValidateError);
        var hasError = !_isEmpty(validationErrors);

        if (!hasError) {
          resolve();
        } else {
          reject(validationErrors.reduce(function (err, v) {
            var key = Object.keys(v)[0];
            if (key) {
              err[key] = v[key];
            }

            return err;
          }, {}));
        }
      });
    });
  };

  this.markAsSaved = function () {
    _this2._dirty = false;
    _this2.removeCache();
  };

  this.selectByIndex = function (index, value) {
    value = value || _this2.props.value;
    index = _isUndefined(index) ? _this2.props.defaultSelectedIndex : index;
    var safeIndex = getSafeSelectedValueIndex(index, value);
    var safeValue = value[safeIndex];

    _this2.setState({
      selectedUUID: _this2.getUUIDFromValue(safeValue),
      showAddComponentOverlay: false
    });
  };

  this.isSelected = function (value) {
    var selectedUUID = _this2.state.selectedUUID;

    return _this2.getUUIDFromValue(value) === selectedUUID;
  };

  this.hasSelected = function () {
    var selectedUUID = _this2.state.selectedUUID;


    return !!selectedUUID;
  };

  this.savePreview = function (instance) {
    if (instance && instance.getDecoratedComponentInstance) {
      instance = instance.getDecoratedComponentInstance();
    }
    _this2.preview = instance;
  };

  this.adjustHeight = function (id) {
    // 不要重复执行
    if (_this2.adjustHeightTimer) {
      clearTimeout(_this2.adjustHeightTimer);
      _this2.adjustHeightTimer = undefined;
    }

    _this2.adjustHeightTimer = setTimeout(function () {
      id = id || _this2.state.selectedUUID;
      if (_this2.preview && _this2.preview.getEditorBoundingBox) {
        var editorBB = _this2.preview.getEditorBoundingBox(id);
        if (!editorBB) {
          return _this2.setState({
            bottomGap: 0
          });
        }

        var previewNode = findDOMNode(_this2.preview);
        var previewBB = previewNode && previewNode.getBoundingClientRect();
        if (!previewBB) {
          return;
        }

        var gap = Math.max(0, editorBB.bottom - previewBB.bottom);
        _this2.setState({
          bottomGap: gap
        });
      }
    }, 0);
  };

  this.onBeforeWindowUnload = function (evt) {
    if (!_this2._dirty) {
      return;
    }

    // 这个字符串其实不会展示给用户
    var confirmLeaveMessage = '页面上有未保存的数据，确定要离开吗？';
    evt.returnValue = confirmLeaveMessage;
    return confirmLeaveMessage;
  };

  this.onRestoreCacheAlertClose = function () {
    _this2.setState({
      showRestoreFromCache: false
    });
  };

  this.restoreCache = function (evt) {
    evt.preventDefault();

    var cachedValue = _this2.readCache();
    if (cachedValue !== storage.NOT_FOUND) {
      _this2.trackValueChange(cachedValue, false);
      _this2.setState({
        showRestoreFromCache: false
      });
      _this2.removeCache();
    }
  };

  this.design = function () {
    return {
      injections: {
        getPreviewProps: function getPreviewProps() /* implementation */{
          // eslint-disable-next-line
          console.warn('Design injections are no longer supported, use `settings` and `onSettingsChange` instead.');

          // this.getPreviewProps = implementation;
        }
      },

      getUUID: _this2.getUUIDFromValue,

      validateComponentValue: _this2.validateComponentValue,

      setValidation: _this2.setValidation,

      markAsSaved: _this2.markAsSaved,

      adjustPreviewHeight: _this2.adjustHeight
    };
  }();
}, _temp);
export { Design as default };


function tagValuesWithUUID(values) {
  values.forEach(function (v) {
    if (!v[UUID_KEY]) {
      v[UUID_KEY] = uuid();
    }
  });
}

/**
 * 从 startIndex 开始往前找到第一个可以选中的值
 * @param {array} value 当前的值
 * @param {array} components 当前可用的组件列表
 * @param {number} startIndex 开始搜索的下标
 */
function findFirstEditableSibling(value, components, startIndex) {
  var loop = function loop(i) {
    var val = value[i];
    var type = val.type;
    var comp = _find(components, function (c) {
      return isExpectedDesignType(c, type);
    });
    if (comp && _defaultTo(comp.editable, true)) {
      return val;
    }
  };

  var valueLength = value.length;
  // 往前找
  for (var i = startIndex; i >= 0 && i < valueLength; i--) {
    var _val = loop(i);
    if (_val) {
      return _val;
    }
  }

  // 往后找
  for (var _i = startIndex + 1; _i < valueLength; _i++) {
    var _val2 = loop(_i);
    if (_val2) {
      return _val2;
    }
  }

  return null;
}

/**
 * 根据当前的值生成一个组件使用计数
 * @param {Array} value Design 当前的值
 * @param {Array} components Design 支持的组件列表
 */
function makeInstanceCountMapFromValue(value, components) {
  var instanceCountMap = new LazyMap(0);

  (value || []).forEach(function (val) {
    var comp = _find(components, function (c) {
      return isExpectedDesignType(c, val.type);
    });
    instanceCountMap.inc(serializeDesignType(comp.type));
  });

  return instanceCountMap;
}

function getSafeSelectedValueIndex(index, value) {
  return Math.min(index, value.length - 1);
}
