// zent/es/select/components/Search.js
import { __extends } from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
var Search = (function (_super) {
    __extends(Search, _super);
    function Search(props) {
        var _this = _super.call(this, props) || this;
        _this.input = null;
        _this.changeHandler = _this.changeHandler.bind(_this);
        _this.focused = false;
        return _this;
    }
    Search.prototype.componentWillReceiveProps = function (nextProps) {
        var _this = this;
        if (!this.focused && nextProps.ready) {
            setTimeout(function () {
                _this.input && _this.input.focus();
            }, 150);
            this.focused = true;
        }
    };
    Search.prototype.changeHandler = function (ev) {
        this.props.onChange(ev.target.value);
    };
    Search.prototype.render = function () {
        var _this = this;
        var _a = this.props, prefixCls = _a.prefixCls, placeholder = _a.placeholder, keyword = _a.keyword;
        return (React.createElement("div", { className: prefixCls + "-search" },
            React.createElement("input", { type: "text", ref: function (input) { return (_this.input = input); }, placeholder: placeholder, className: prefixCls + "-filter", value: keyword, onChange: this.changeHandler })));
    };
    return Search;
}(PureComponent));
export default Search;
