// zent/es/sortable/Sortable.js
import { __assign, __extends, __rest } from "tslib";
import * as React from 'react';
import { Component } from 'react';
import cx from 'classnames';
import * as sortableJS from 'sortablejs';
import reorder from '../utils/reorder';
var Sortable = (function (_super) {
    __extends(Sortable, _super);
    function Sortable() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.initSortable = function (instance) {
            var _a = _this.props, prefix = _a.prefix, onMove = _a.onMove, onEnd = _a.onEnd, onChange = _a.onChange, filterClass = _a.filterClass, children = _a.children, rest = __rest(_a, ["prefix", "onMove", "onEnd", "onChange", "filterClass", "children"]);
            if (!instance) {
                return;
            }
            var sortableOptions = __assign({ filter: filterClass ? "." + filterClass : '', ghostClass: prefix + "-ghost", chosenClass: prefix + "-chosen", dragClass: prefix + "-drag", fallbackClass: prefix + "-fallback", onMove: function (e) {
                    if (onMove) {
                        return onMove(e);
                    }
                    return e.related.className !== filterClass;
                }, onEnd: function (e) {
                    var items = _this.props.items;
                    onEnd && onEnd(e);
                    if (!items) {
                        return;
                    }
                    var oldIndex = e.oldIndex, newIndex = e.newIndex;
                    var newItems = reorder(items, oldIndex, newIndex);
                    onChange && onChange(newItems);
                } }, rest);
            _this.sortable = sortableJS.create(instance, sortableOptions);
        };
        return _this;
    }
    Sortable.prototype.componentWillUnmount = function () {
        if (this.sortable) {
            this.sortable.destroy();
            this.sortable = null;
        }
    };
    Sortable.prototype.render = function () {
        var _this = this;
        var _a = this.props, prefix = _a.prefix, className = _a.className, children = _a.children, tag = _a.tag;
        var classString = cx(prefix + "-sortable", className);
        var Com = tag;
        return (React.createElement(Com, { ref: function (instance) { return _this.initSortable(instance); }, className: classString }, children));
    };
    Sortable.defaultProps = {
        prefix: 'zent',
        tag: 'div',
    };
    return Sortable;
}(Component));
export { Sortable };
export default Sortable;
