// formulr/lib/models/array.js
import { __extends, __spreadArrays } from "tslib";
import { BehaviorSubject } from 'rxjs';
import { BasicModel, isModel } from './basic';
import { ValidateOption } from '../validate';
import { ModelRef, isModelRef } from './ref';
import { Some, or } from '../maybe';
var FIELD_ARRAY = Symbol('field-array');
var FieldArrayModel = /** @class */ (function (_super) {
    __extends(FieldArrayModel, _super);
    /** @internal */
    function FieldArrayModel(childBuilder, defaultValue) {
        var _this = _super.call(this) || this;
        _this.defaultValue = defaultValue;
        _this.childFactory = childBuilder
            ? function (defaultValue) { return childBuilder.build(Some(defaultValue)); }
            : function (defaultValue) { return new ModelRef(null, Some(defaultValue), _this); };
        var children = _this.defaultValue.map(_this.childFactory);
        _this.children$ = new BehaviorSubject(children);
        return _this;
    }
    FieldArrayModel.prototype.reset = function () {
        var children = or(this.initialValue, this.defaultValue).map(this.childFactory);
        this.children$.next(children);
    };
    FieldArrayModel.prototype.clear = function () {
        this.initialValue = undefined;
        var children = this.defaultValue.map(this.childFactory);
        this.children$.next(children);
    };
    Object.defineProperty(FieldArrayModel.prototype, "children", {
        get: function () {
            return this.children$.getValue();
        },
        enumerable: true,
        configurable: true
    });
    FieldArrayModel.prototype.valid = function () {
        if (this.error$.getValue() !== null) {
            return false;
        }
        var children = this.children$.getValue();
        for (var i = 0; i < children.length; i += 1) {
            var child = children[i];
            if (isModelRef(child)) {
                var model = child.getModel();
                if (!model || !model.valid()) {
                    return false;
                }
            }
            else if (isModel(child) && !child.valid()) {
                return false;
            }
        }
        return true;
    };
    FieldArrayModel.prototype.getRawValue = function () {
        return this.children$.getValue().map(function (child) {
            if (isModelRef(child)) {
                var model = child.getModel();
                return model ? model.getRawValue() : null;
            }
            else if (isModel(child)) {
                return child.getRawValue();
            }
            return null;
        });
    };
    FieldArrayModel.prototype.getSubmitValue = function () {
        return this.children$.getValue().map(function (child) {
            if (isModelRef(child)) {
                var model = child.getModel();
                return model ? model.getSubmitValue() : null;
            }
            else if (isModel(child)) {
                return child.getSubmitValue();
            }
            return null;
        });
    };
    FieldArrayModel.prototype.patchValue = function (value) {
        var children = this.children$.getValue();
        for (var i = 0; i < value.length; i += 1) {
            if (i >= children.length) {
                break;
            }
            var item = value[i];
            var model = children[i];
            if (isModelRef(model)) {
                var m = model.getModel();
                m && m.patchValue(item);
            }
            else if (isModel(model)) {
                model.patchValue(item);
            }
        }
        if (value.length <= children.length) {
            this.splice(value.length, children.length - value.length);
            return;
        }
        for (var i = children.length; i < value.length; i += 1) {
            var item = value[i];
            this.push(item);
        }
    };
    FieldArrayModel.prototype.initialize = function (values) {
        this.initialValue = Some(values);
        this.children$.next(values.map(this.childFactory));
    };
    FieldArrayModel.prototype.push = function () {
        var items = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            items[_i] = arguments[_i];
        }
        var nextChildren = this.children$.getValue().concat(items.map(this.childFactory));
        this.children$.next(nextChildren);
    };
    FieldArrayModel.prototype.pop = function () {
        var children = this.children$.getValue().slice();
        var child = children.pop();
        this.children$.next(children);
        return child;
    };
    FieldArrayModel.prototype.shift = function () {
        var children = this.children$.getValue().slice();
        var child = children.shift();
        this.children$.next(children);
        return child;
    };
    FieldArrayModel.prototype.unshift = function () {
        var items = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            items[_i] = arguments[_i];
        }
        var nextChildren = items.map(this.childFactory).concat(this.children$.getValue());
        this.children$.next(nextChildren);
    };
    FieldArrayModel.prototype.splice = function (start, deleteCount) {
        if (deleteCount === void 0) { deleteCount = 0; }
        var items = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            items[_i - 2] = arguments[_i];
        }
        var children = this.children$.getValue().slice();
        var ret = children.splice.apply(children, __spreadArrays([start, deleteCount], items.map(this.childFactory)));
        this.children$.next(children);
        return ret;
    };
    FieldArrayModel.prototype.validate = function (option) {
        if (option === void 0) { option = ValidateOption.Default; }
        if (option & ValidateOption.IncludeChildrenRecursively) {
            return Promise.all(this.children$
                .getValue()
                .map(function (it) { return it.validate(option); })
                .concat(this.triggerValidate(option)));
        }
        return this.triggerValidate(option);
    };
    FieldArrayModel.prototype.pristine = function () {
        var children = this.children$.getValue();
        for (var i = 0; i < children.length; i += 1) {
            var child = children[i];
            if (child.dirty()) {
                return false;
            }
        }
        return true;
    };
    FieldArrayModel.prototype.dirty = function () {
        return !this.pristine();
    };
    FieldArrayModel.prototype.touched = function () {
        var children = this.children$.getValue();
        for (var i = 0; i < children.length; i += 1) {
            var child = children[i];
            if (child.touched()) {
                return true;
            }
        }
        return false;
    };
    return FieldArrayModel;
}(BasicModel));
FieldArrayModel.prototype[FIELD_ARRAY] = true;
function isFieldArrayModel(maybeModel) {
    return !!(maybeModel && maybeModel[FIELD_ARRAY]);
}
export { FieldArrayModel, isFieldArrayModel };
